interpreter/ruby/ruby.go

```diff
@@ -628,57 +727,445 @@ func (r *rubyInstance) getRubyLineNo(iseqBody libpf.Address, pc uint64) (uint32,
 	return lineNo, nil
 }
 
-func (r *rubyInstance) Symbolize(frame *host.Frame, frames *libpf.Frames) error {
-	if !frame.Type.IsInterpType(libpf.Ruby) {
-		return interpreter.ErrMismatchInterpreterType
+func (r *rubyInstance) readClassName(classAddr libpf.Address) (libpf.String, bool, error) {
+	var classPath libpf.String
+	var classpathPtr libpf.Address
+	var singleton bool
+	var err error
+
+	classFlags := r.rm.Ptr(classAddr)
+	classMask := classFlags & rubyTMask
+
+	switch classMask {
+	case rubyTClass, rubyTModule:
+		classpathPtr = r.rm.Ptr(classAddr + libpf.Address(r.r.vmStructs.rclass_and_rb_classext_t.classext+r.r.vmStructs.rb_classext_struct.classpath))
+
+		// Should also check if it is a singleton
+		// https://github.com/ruby/ruby/blob/b627532/vm_backtrace.c#L1934-L1937
+		// https://github.com/ruby/ruby/blob/b627532/internal/class.h#L528
+
+		if classFlags&RUBY_FL_SINGLETON != 0 {
+			log.Debugf("Got singleton class")
```

Comment:
todo: delete me and clean up these comments, add a comment about anonymous frame handling, link to rbspy source for this and upstream ruby source.